### ISSUE TITLE
ParadoxicalCondition false positive

### ISSUE THREAD

**Issue by hxv** [vimeo/psalm#2821]
```php
$onlyInvalid = (bool) rand(0, 1);
$onlyValid = (bool) rand(0, 1);
$valid = (bool) rand(0, 1);

if ($onlyInvalid && $onlyValid) {
  return;
}

if (($onlyInvalid && $valid) || ($onlyValid && !$valid)) {
}
```
> Psalm output (using commit 422bd25):
>
> ERROR: ParadoxicalCondition - 11:5 - Encountered a paradox when evaluating the conditionals (!$onlyInvalid || !$onlyValid) and (!$onlyInvalid || !$onlyValid)

https://psalm.dev/r/41f376addb

That doesn't look like a paradox to me.

Weirdly enough everything seems fine after changing first condition to:
```php
if ($onlyInvalid && $onlyValid) {
```
https://psalm.dev/r/82b5c9ba21

**Comment by muglug** [comment]
I _think_ this is the first time someone's found a foundational issue with Psalm's boolean simplification logic.

The problem is that Psalm converts

```
($a && $c) || ($b && !$c)
```
(disjunctive normal form)

into

```
($a || $b) && ($a || !$c) && ($c || $b) && ($c || !$c)
```
(conjunctive normal form)

which Psalm can simplify to

```
($a || $b) && ($a || !$c) && ($c || $b)
```

but _no further_.

In fact the first term (and the one that's causing the bug above) can in fact be removed, as `$a || $b` is true for all subsequent terms in the clause.

Improving this simplification would, I think, make the boolean simplification algorithm much more expensive, unfortunately.

Please either suppress it or change the conditional to its conjunctive normal form: https://psalm.dev/r/40ed58cbf2
